### PR TITLE
Take a backup before applying pod priorities to bss-etcd and add some delays for the etcd operator to reconcile

### PR DESCRIPTION
**Pull request by bklei**
## Summary and Scope

Take a backup of bss before applying pod priorities, wait for deployments/statefulsets/daemonsets to roll out with their changes, as well as introduce a delay after re-adding a node to give the etcd operator a chance to complete a reconcile.

## Issues and Related PRs

* Resolves [CASMINST-4597](https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-4597)

## Testing

```
surtur-ncn-m001-pit:/usr/share/doc/csm/upgrade/1.0.11 # ./scripts/upgrade/add_pod_priority.sh
Found 3 running pods in cray-bss-etcd etcd cluster
Found 3 ready members in cray-bss-etcd etcd cluster
Found status of Running for cray-bss-etcd etcd cluster
Sleeping for thirty seconds to let the etcd operator reconcile
Creating csm-high-priority-service pod priority class
priorityclass.scheduling.k8s.io/csm-high-priority-service configured

Patching cray-bss deployment in services namespace
deployment.apps/cray-bss patched
Waiting for deployment spec update to be observed...
Waiting for deployment "cray-bss" rollout to finish: 2 out of 3 new replicas have been updated...
Waiting for deployment "cray-bss" rollout to finish: 2 out of 3 new replicas have been updated...
Waiting for deployment "cray-bss" rollout to finish: 1 old replicas are pending termination...
Waiting for deployment "cray-bss" rollout to finish: 1 old replicas are pending termination...
deployment "cray-bss" successfully rolled out

Patching cray-keycloak-gatekeeper-ingress deployment in services namespace
deployment.apps/cray-keycloak-gatekeeper-ingress patched (no change)
deployment "cray-keycloak-gatekeeper-ingress" successfully rolled out

Patching cray-postgres-operator deployment in services namespace
deployment.apps/cray-postgres-operator patched (no change)
deployment "cray-postgres-operator" successfully rolled out

Patching cray-postgres-operator-postgres-operator-ui deployment in services namespace
deployment.apps/cray-postgres-operator-postgres-operator-ui patched (no change)
deployment "cray-postgres-operator-postgres-operator-ui" successfully rolled out

Patching istio-operator deployment in istio-operator namespace
deployment.apps/istio-operator patched (no change)
deployment "istio-operator" successfully rolled out

Patching istio-ingressgateway deployment in istio-system namespace
deployment.apps/istio-ingressgateway patched (no change)
deployment "istio-ingressgateway" successfully rolled out

Patching istio-ingressgateway-hmn deployment in istio-system namespace
deployment.apps/istio-ingressgateway-hmn patched (no change)
deployment "istio-ingressgateway-hmn" successfully rolled out

Patching istiod deployment in istio-system namespace
deployment.apps/istiod patched (no change)
deployment "istiod" successfully rolled out

Patching spire-jwks deployment in spire namespace
deployment.apps/spire-jwks patched (no change)
deployment "spire-jwks" successfully rolled out

Patching cray-ipxe deployment in services namespace
deployment.apps/cray-ipxe patched (no change)
deployment "cray-ipxe" successfully rolled out

Patching cray-tftp deployment in services namespace
deployment.apps/cray-tftp patched (no change)
deployment "cray-tftp" successfully rolled out

Patching cray-dhcp-kea deployment in services namespace
deployment.apps/cray-dhcp-kea patched (no change)
deployment "cray-dhcp-kea" successfully rolled out

Patching cray-dns-unbound deployment in services namespace
deployment.apps/cray-dns-unbound patched (no change)
deployment "cray-dns-unbound" successfully rolled out

Patching nexus deployment in nexus namespace
deployment.apps/nexus patched (no change)
deployment "nexus" successfully rolled out

Patching cray-opa deployment in opa namespace
deployment.apps/cray-opa patched (no change)
deployment "cray-opa" successfully rolled out

Patching cray-smd deployment in services namespace
deployment.apps/cray-smd patched (no change)
deployment "cray-smd" successfully rolled out

Patching cray-ceph-csi-cephfs-provisioner deployment in default namespace
deployment.apps/cray-ceph-csi-cephfs-provisioner patched (no change)
deployment "cray-ceph-csi-cephfs-provisioner" successfully rolled out

Patching cray-ceph-csi-rbd-provisioner deployment in default namespace
deployment.apps/cray-ceph-csi-rbd-provisioner patched (no change)
deployment "cray-ceph-csi-rbd-provisioner" successfully rolled out

Patching cray-metallb-speaker daemonset in metallb-system namespace
daemonset.apps/cray-metallb-speaker patched
Waiting for daemon set "cray-metallb-speaker" rollout to finish: 1 out of 3 new pods have been updated...
Waiting for daemon set "cray-metallb-speaker" rollout to finish: 1 out of 3 new pods have been updated...
Waiting for daemon set "cray-metallb-speaker" rollout to finish: 2 out of 3 new pods have been updated...
Waiting for daemon set "cray-metallb-speaker" rollout to finish: 2 out of 3 new pods have been updated...
Waiting for daemon set "cray-metallb-speaker" rollout to finish: 2 of 3 updated pods are available...
daemon set "cray-metallb-speaker" successfully rolled out

Patching cray-keycloak statefulset in services namespace
statefulset.apps/cray-keycloak patched
Waiting for 1 pods to be ready...
Waiting for 1 pods to be ready...
Waiting for 2 pods to be ready...
Waiting for 2 pods to be ready...
Waiting for 1 pods to be ready...
statefulset rolling update complete 1 pods at revision cray-keycloak-564845dcf4...

Patching keycloak-postgres statefulset in services namespace
statefulset.apps/keycloak-postgres patched (no change)
statefulset rolling update complete 3 pods at revision keycloak-postgres-5699855b47...

Patching spire-postgres statefulset in spire namespace
statefulset.apps/spire-postgres patched (no change)
statefulset rolling update complete 3 pods at revision spire-postgres-7cb45cd758...

Patching spire-server statefulset in spire namespace
statefulset.apps/spire-server patched (no change)
partitioned roll out complete: 3 new pods have been updated...

Patching cray-smd-postgres statefulset in services namespace
statefulset.apps/cray-smd-postgres patched (no change)
statefulset rolling update complete 3 pods at revision cray-smd-postgres-dc4d5d97d...

Creating a backup of cray-bss etcdcluster prior to restarting the cluster
etcdbackup.etcd.database.coreos.com/cray-bss-etcd-cluster-manual-backup-14738 created
Patching cray-bss-etcd etcdcluster in services namespace
etcdcluster.etcd.database.coreos.com/cray-bss-etcd patched (no change)

pod "cray-bss-etcd-cfqz9nbdvr" deleted
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Found 3 running pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 ready members and 'Running' state for cray-bss-etcd etcd cluster
Found 3 ready members in cray-bss-etcd etcd cluster
Found status of Running for cray-bss-etcd etcd cluster
Sleeping for thirty seconds to let the etcd operator reconcile
pod "cray-bss-etcd-dzlwz76f5w" deleted
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Found 3 running pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 ready members and 'Running' state for cray-bss-etcd etcd cluster
Found 3 ready members in cray-bss-etcd etcd cluster
Found status of Running for cray-bss-etcd etcd cluster
Sleeping for thirty seconds to let the etcd operator reconcile
pod "cray-bss-etcd-vf5njv26cl" deleted
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Sleeping for ten seconds waiting for 3 pods in cray-bss-etcd etcd cluster
Found 3 running pods in cray-bss-etcd etcd cluster
Found 3 ready members in cray-bss-etcd etcd cluster
Found status of Running for cray-bss-etcd etcd cluster
Sleeping for thirty seconds to let the etcd operator reconcile
```

### Tested on:

  * `surtur`

### Test description:

Ran the changed script on surtur

## Risks and Mitigations

Low

## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: install/install_csm_services.md [hunk 1]
```diff
@@ -365,6 +365,17 @@ deployment.apps/istio-ingressgateway patched
 
 After running the `add_pod_priority.sh` script, the affected pods will be restarted as the pod priority class is applied to them.
 
+> **`NOTE`** If the script doesn't finish and ends up looping on this message
+> longer that several minutes:
```
Comment (cdelatte-hpe) [inline install/install_csm_services.md:376]:
```suggestion
> longer than several minutes:
```

**Review by cdelatte-hpe — CHANGES_REQUESTED** [review]

**Review by cdelatte-hpe — APPROVED** [review]

**Review by rgross727 — APPROVED** [review]

File: upgrade/1.0.1/scripts/upgrade/add_pod_priority.sh [hunk 2]
```diff
@@ -147,10 +151,14 @@ for statefulset in $STATEFULSETS; do
   if [ ! -z "$ns" ]; then
     echo "Patching $statefulset statefulset in $ns namespace"
     kubectl -n $ns patch statefulset $statefulset --type merge -p '{"spec": {"template": {"spec": {"priorityClassName": "csm-high-priority-service"}}}}'
+    kubectl rollout status statefulset -n $ns $statefulset
     echo ""
   fi
 done
 
+echo "Creating a backup of cray-bss etcdcluster prior to restarting the cluster"
+kubectl exec -it -n operators $(kubectl get pod -n operators | grep etcd-backup-restore | head -1 | awk '{print $1}') -c util -- create_backup cray-bss pod-priority-backup-$(date "+%D-%T") | grep -v 'unknown operand'
```
Comment (dborman-hpe) [inline upgrade/1.0.1/scripts/upgrade/add_pod_priority.sh:160]: An alternate way to get the pod name using just kubectl:
```
kubectl get pod -n operators -l app.kubernetes.io/name=cray-etcd-backup -o jsonpath='{.items[0].metadata.name}'
```

Comment (dborman-hpe) [reply upgrade/1.0.1/scripts/upgrade/add_pod_priority.sh:160]: Note: this is just a suggestion, no change needed.

Comment (bklei) [reply upgrade/1.0.1/scripts/upgrade/add_pod_priority.sh:160]: good suggestion, will fix

**Pull request merged**